### ISSUE TITLE
Add content to `The GitHub workflow` wiki page

### ISSUE THREAD

**Issue by KevinTran000** [LibreFoodPantry/BEAR-Necessities-Market#32]
Add content to `The GitHub workflow` wiki page.

List the standard workflow that should be used for contributing to this repo.

**Comment by StoneyJackson** [comment]
Agreed. Here's how I propose working.

## Basic workflow: fork-branch-pull-request

1. Fork upstream
2. Create a branch in fork
3. Work in branch on fork
4. Issue pull-request from branch in fork to master in upstream

## Early pull-requests for discussions and awareness

So that others can more easily track your progress and provide early feedback, consider issuing a pull-request early, before you are done. GitHub even has [draft pull-requests](https://github.blog/2019-02-14-introducing-draft-pull-requests/) to support this practice.

## Why branch?

Technically the branch in your fork is optional. From the project's perspective it doesn't matter. However, it has several advantages from your perspective.

Working in a branch means you will never change master directly. In fact, all changes to master will come from upstream when you `git checkout master; git pull upstream master`. The only time your changes are merged into master is when a maintainer merges your pull-request into master in upstream.

The separation between upstream's master and your branch is nice because you always know that master represents what is in upstream, and your branch represents your work. This also gives you control over when to merge upstream into your branch. You wouldn't want to merge when you are in the middle of fixing something, because then it would be harder to know if the changes in upstream broke something in your code or vice versa.

Last, if development stalls on your branch, or you need to work on a different feature or bug, you can start a new branch (off of master; always off of master) and work on the new feature/bug without losing your other work. You can't do this if you work directly on master without some complicated git wizardry.

## The devil's in the details

Details of this workflow are described here: https://github.com/foss2serve/github-workflow-activity/blob/master/activity/contribution-workflow.md .

As maintainers, you may want to not squash a PR when merging, that way all attributions in commit messages are preserved.

## How to do this in teams?

The workflow described above is designed for individuals working on a single project. So how do you manage a team working on a single project? Below is one way.

1. Maintainers create a branch for each feature that is being worked on in this sprint, and configure TravisCI to check pull-requests made into these feature branches.
2. Everyone working on a particular feature treats the feature-branch as master in the workflow above. In other words, everywhere the workflow refers to master, they should use the feature-branch; and everywhere the workflow refers to the feature-branch they should use an individual-feature-branch.

This model minimizes setup, allows anyone to contribute to any effort, and leverages TravisCI.

What still needs to be worked out is how to coordinate teams with issues, projects (Trello-like board), milestones, or whatever.

**Comment by StoneyJackson** [comment]
I'm not sure I'm entirely in love with what I proposed above, so suggestions and feedback is welcome.

**Comment by StoneyJackson** [comment]
Another way to go:

1. A team member forks upstream, creates the feature branch, issues draft pull-request. This new fork is the team's fork.
2. That same team member protects the master branch in the team's fork so that it cannot be pushed to.
3. That same team member adds other team members as contributors.
4. Team members clone the team's fork directly.
5. Team members work on the feature branch directly as follows

```
git checkout the-feature-branch
[make a small change or two]
git add .
git commit -m "small change or two"
git pull --rebase origin the-feature-branch
git push origin the-feature-branch 
```

Note the `git pull --rebase origin the-feature-branch`. This step pulls any changes other team members have added into the individual's clone and re-applies the individual's changes on top of them (rebase). There may be conflicts at this point which the individual would have to resolve.

Team members make small little changes that are reflected in the pull request immediately. So you are still getting all the benefits of TravisCI. Also, the team can use the issue tracker and a project board to organize their team.

I'm like this more than my original post.

**Comment by StoneyJackson** [comment]
To summarize...

- There is upstream project on GitHub. The upstream project has a project board to organize and coordinate teams working on features.
- Each team forks upstream into a team-fork on GitHub.
- All team members are given commit privileges to their team's fork.
- The master branch in all team's forks is protected so commits cannot be directly pushed into master.
- Each team uses the issue tracker and project board of their team's fork to break down a feature into tasks and to organize and coordinate their team.
- Each team creates a feature-branch for each feature they work on, and create a draft-pull-request back to upstream's master.
- Each team member clones their team's fork and checks out the feature-branch that they will work on, and commits and pushes small frequent changes.
- The pull-request for each feature can be observed and monitored for TravisCI results and discussions between teams and maintainers.

Basically the team is acting as a collective individual in the original fork-branch-pull-request workflow. Synchronization may still be necessary as features are merged into upstream's master.

This feels good! What do others think?

**Comment by KevinTran000** [comment]
I like this idea very much. :+1:
I had the same conversation with a few other people last week and we reached the same conclusion as well, a team fork.

**Comment by StoneyJackson** [comment]
https://github.com/LibreFoodPantry/BEAR-Necessities-Market/wiki/The-GitHub-workflow
